gwcs/wcs.py

```diff
@@ -634,3 +641,211 @@ def fix_inputs(self, fixed):
+        U = (cdmat[1, 1] * undist_x - cdmat[0, 1] * undist_y) / det
+        V = (-cdmat[1, 0] * undist_x + cdmat[0, 0] * undist_y) / det
+        fit_inv_poly_u, fit_inv_poly_v, max_inv_resid = _fit_2D_poly(degree,
+                                                            max_inv_error,
+                                                            U, V, u - U, v - V,
+                                                            verbose=verbose)
+        pdegree = fit_poly_x.degree
+        if pdegree > 1:
+            hdr['a_order'] = pdegree
+            hdr['b_order'] = pdegree
+            ipdegree = fit_inv_poly_u.degree
+            hdr['ap_order'] = ipdegree
+            hdr['bp_order'] = ipdegree
+            cd, sip_poly_x, sip_poly_y = _reform_poly_coefficients(fit_poly_x, fit_poly_y)
+            _store_2D_coefficients(hdr, sip_poly_x, 'A')
+            _store_2D_coefficients(hdr, sip_poly_y, 'B')
+            _store_2D_coefficients(hdr, fit_inv_poly_u, 'AP', keeplinear=True)
+            _store_2D_coefficients(hdr, fit_inv_poly_v, 'BP', keeplinear=True)
+            hdr['sipmxerr'] = (max_resid, 'Maximum difference from the GWCS model SIP is fit to.')
+            hdr['sipiverr'] = (max_inv_resid, 'Maximum difference for the inverse transform')
+
+        else:
+            hdr['ctype1'] = 'RA---TAN'
+            hdr['ctype2'] = 'DEC--TAN'
+
+        hdr['cd1_1'] = cd[0][0]
+        hdr['cd1_2'] = cd[0][1]
+        hdr['cd2_1'] = cd[1][0]
+        hdr['cd2_2'] = cd[1][1]
+        return hdr
+
+
+def _fit_2D_poly(degree, max_error, u, v, x, y, verbose=False):
+    """
+    Fit a pair of ordinary 2D polynomial to the supplied inputs (u, v) and
+    outputs (x, y)
+    """
+    llsqfitter = LinearLSQFitter()
+    # The case of one pass with the specified polynomial degree
+    if degree is None:
+        deglist = range(1,10)
+    else:
+        deglist = [degree]
+    for deg in deglist:
+        poly_x = Polynomial2D(degree=deg)
+        poly_y = Polynomial2D(degree=deg)
+        fit_poly_x = llsqfitter(poly_x, u, v, x)
+        fit_poly_y = llsqfitter(poly_y, u, v, y)
+        max_resid = _compute_distance_residual(x, y,
+                                               fit_poly_x(u, v), fit_poly_y(u, v))
+        if verbose:
+            print(f'Degree = {degree}, max_resid = {max_resid}')
+        if max_error is not None and max_resid < max_error:
+            if verbose:
+                print('terminating condition met')
+            break
+        if degree is None and deg == 9:
+            raise RuntimeError("Unable to achieve required fit accuracy with SIP polynomials")
+    return fit_poly_x, fit_poly_y, max_resid
+
+
+def _compute_distance_residual(undist_x, undist_y, fit_poly_x, fit_poly_y):
+    """
+    Compute the distance residuals and return the rms and maximum values.
+    """
+    dist = np.sqrt((undist_x - fit_poly_x)**2 + (undist_y - fit_poly_y)**2)
+    max_resid = dist.max()
```

Comment:
I think it would be more efficient to compute `sqrt()` just once on the `max` value:
```suggestion
    max_resid = np.sqrt(((undist_x - fit_poly_x)**2 + (undist_y - fit_poly_y)**2).max())
```